### DISCUSSION TITLE
Auto Inertia Computation

### DISCUSSION THREAD

**Discussion by rohit-kumar-j** [google-deepmind/mujoco#318] [General]
By default, does mujoco support auto inertia computation from an STL file loaded as a body? Like a [hollow vase](https://cults3d.com/en/3d-model/home/vase-006)? (assuming uniform density)

**Comment by yuvaltassa** [comment]
Quoting the [documentation](https://mujoco.readthedocs.io/en/latest/XMLreference.html?highlight=pyramids#asset-mesh):

> The inertial computation mentioned above is part of an algorithm used not only to center and align the mesh, but also to infer the mass and inertia of the body to which it is attached. This is done by computing the centroid of the triangle faces, connecting each face with the centroid to form a triangular pyramid, computing the mass and inertia of all pyramids and accumulating them. This algorithm comes from Astronomy where it is used to estimate [inertial](https://mujoco.readthedocs.io/en/latest/XMLreference.html?highlight=pyramids#inertial) properties of asteroids. It is exact for convex meshes but is not always exact for non-convex meshes; indeed no algorithm can be exact when the notion of interior is ill-defined. Thus for non-convex models designed in CAD software (which usually knows what the interior is) it is better to ask that software to compute the inertial properties of the body and enter them in the MJCF file explicitly via the inertial element.

So hollow objects, like a vase, would be likely badly approximated. I'm actually quite curious to know how bad the approximation is. If the algorithm, implemented [here](https://github.com/deepmind/mujoco/blob/main/src/user/user_mesh.cc#L938) uses a **signed volume**, one would expect a hollow convex shape, like a hollow cube, to produce exact results; but I have not tested this. If you feel like testing what inertia you get for a centred hollow-convex shape of known inertia (the inertia of a hollow box is just the difference of inertias of the uniform box and the inertia of box taken from the "hole"), that would be grand.

**Comment by rohit-kumar-j** [comment]
I tested it using this [Rotational Inertia: The Race Between a Ring and a Disc](https://www.youtube.com/watch?v=CHQOctEvtTY&ab_channel=NorthCarolinaSchoolofScienceandMathematics). Seems like the hollow object's inertia auto-generated correctly. But I haven't checked the distance vs time, etc to actually verify if it is correct. I chose the disc vs ring since there is a direct observable difference between the two as shown in the video. If you have another test in mind please suggest it, I would gladly test it out.

For this expt:
* The masses are 1.0 under `geom` for both the bodies.
* The outer dimensions are the same for both the disc and the ring
* Both the bodies are meshes, inertias are not mentioned in the xml

I have a question regarding the expt:
Is the diameter of the disc too well defined to verify how sensitive the algorithm is?
> >It is exact for convex meshes but is not always exact for non-convex meshes; indeed no algorithm can be exact when the notion of interior is ill-defined.

<details>
    <Summary>Xml  and Meshes</Summary>

Meshes:
[rolling_rings.zip](https://github.com/deepmind/mujoco/files/8839413/rolling_rings.zip)

```xml
<mujoco>
    <option gravity="0 0 -9.81" timestep="0.002" />
    <asset>
        <mesh name="Ring" file="./hollow_ring.stl" scale=".001 .001 .001" />
        <mesh name="Disc" file="./solid_ring.stl" scale=".001 .001 .001" />
    </asset>
    <worldbody>
        <geom type="plane" size="10 10 0.1" pos="0 0 -0.5" euler="10 0 0" />
        <light diffuse=".5 .5 .5" pos="0 0 3" dir="0 0 -1" />

        <body name="Ring" pos="1 0 0" euler="0 90 0">
            <joint type="free" />
            <geom type="mesh" mesh="Ring" rgba=".3 .7 .1 1" mass="1.0" />
        </body>
        <body name="Disc" pos="0 0 0" euler="0 90 0">
            <joint type="free" />
            <geom type="mesh" mesh="Disc" rgba=".6 .7 .5 1" mass="1.0" />
        </body>
    </worldbody>
</mujoco>
```

</details>

<details>
    <Summary>Python file</Summary>

```python
import mujoco
from mujoco_viewer import MujocoViewer # pip install mujoco-python-viewer

model = mujoco.MjModel.from_xml_path(
    "assets/misc/rolling_rings/rolling_rings.xml"
)
data = mujoco.MjData(model)
viewer = MujocoViewer(model, data)

while True:
    mujoco.mj_step(model, data)
    viewer.render()
```
</details>

---
## Test

<details>
    <Summary>Video</Summary>

https://user-images.githubusercontent.com/37873142/172037848-c94db584-c7a9-44b7-bf31-41ba8b542ff4.mp4

</details>

<details>
    <Summary>Bodies</Summary>
<img src="https://user-images.githubusercontent.com/37873142/172037726-b9ec3aa7-a395-4855-b4f0-aa89fc860be1.png" width="450">
</details>

<details>
    <Summary>Wire frame</Summary>
<img src="https://user-images.githubusercontent.com/37873142/172037865-bc254256-09ce-4a6f-838c-fd5eb60f1c45.png" width="450">
</details>

### Directly observable differences here:

<details>
    <Summary>Inertias</Summary>
<img src="https://user-images.githubusercontent.com/37873142/172038106-e11789b9-d855-45a6-9b78-68aa3b6a57be.png" width="450">
</details>

**Comment by yuvaltassa** [comment]
In order for this to be precise, we need to use shapes which are not polyhedral approximations of smooth geometries.

You can do the following:
1. Create body with a primitive box of `size=".05 .05 .05"`, i.e. a 10cm^3 box. With the default density of 1000 (kg/m^3) this would be a 1 Kg box. See that the inertia is what you expect.
2. Create body with a primitive box of `size=".04 .04 .04"`, i.e. an 8cm^3 box. See that the inertia is what you expect.
3. Make a mesh in a CAD program that is a 10cm^3 box with a centered 8cm^3 hole, report the computed inertia in MuJoCo, as it compares to the difference of the two inertias above.

Thanks for this!

As a side note, an easy feature to add would be a `boundarymass` attribute (defaults to 0, bounded between 0 and 1) to geoms, which would specify the proportion of mass residing on the boundary of the shape. Setting this to 1 would likely lead to a much better approximation for meshes with thin shells.

**Reply by rohit-kumar-j** [reply]
Can we get the inertia of a body like `getInertiaOfBody(body_id)` `getInertiaOfGeom(geom_id)` for the boxes? I can compare it with a cad software's inertia computation.

**Reply by yuvaltassa** [reply]
The way to do that is exactly what I described above. After loading a model in `simulate` you can hit "print model" which will save a text file with all the fields, including mass/inertia.

In any case, I've since done this test myself and concluded that the negative space is not accounted for.

We have an internal work item to improve this computation. If you wish you can open a GH issue and we will update there. Please follow the [guidelines](https://github.com/deepmind/mujoco/blob/main/CONTRIBUTING.md#suggesting-enhancements) if you do that.
